addons/mass_mailing/controllers/main.py

```diff
@@ -55,3 +87,52 @@ def full_url_redirect(self, code, stat_id, **post):
 
         request.env['link.tracker.click'].add_click(code, request.httprequest.remote_addr, country_code, stat_id=stat_id)
         return werkzeug.utils.redirect(request.env['link.tracker'].get_url_from_code(code), 301)
+
+    @http.route('/mailing/blacklist/check', type='json', auth='none')
+    def blacklist_check(self, email):
+        if email:
+            record = request.env['mail.blacklist'].sudo().with_context(active_test=False).search([('email', '=ilike', email)])
+            if record['active']:
+                return True
+            return False
+        return 'error'
+
+    @http.route('/mailing/blacklist/add', type='json', auth='none')
+    def blacklist_add(self, mailing_id, res_id, email, token):
+        mailing = request.env['mail.mass_mailing'].sudo().browse(mailing_id)
+        if mailing._unsubscribe_token(res_id, email) != token:
+            return 'unauthorized'
+        if email:
+            recipient_name = 'lead' if mailing.mailing_model_real == 'crm.lead' \
```

Comment:
If you are talking about `crm.lead` this code should in `crm`or a bridge between mass_mailing and crm. Or crm should be in dependance of mass_mailling ...